content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+
+enum Optional<Wrapped> {
+    case none
+    case some(Wrapped)
+}
+```
+
+Pretty cool, right? The Swift compiler makes all nullable values `Optional` under the hood.
+In other words, `?` and `nil` are just syntax sugar literals for
+[`enum Optional`](https://developer.apple.com/documentation/swift/optional),
+which is extremely similar to a Kotlin `sealed class Optional`
+like [Koptional](https://github.com/gojuno/koptional).
+Swift is open source so you can take a look at
+[the `Optional` implementation](https://github.com/apple/swift/blob/master/stdlib/public/core/Optional.swift)
+yourself since I’ve simplified it drastacially at the code sample above.
+
+Unfortunately, it is not really possible to change Kotlin behave the same way.
+Apple uses a bridging mechanism to connect Objective-C and Swift binaries, when Kotlin
+uses the same bytecode as Java. To make a simpler mental picture,
+imagine Objective-C and Swift being connected side-by-side and
+Kotlin and Java as a stack, where Kotlin is on top.
+I presume it would be pretty challenging to provide a proper compatibility with Java,
+transforming all nullable Kotlin values to `Optional` and vice-versa,
+especially for such tight areas like Java reflection.
+
+# In Retrospect
+
+Initially, I wanted to name this article as “Kotlin: the Problem with Java”.
+But it is a much wider topic for discussion and maybe I’ll post more about that in the future.
+
+Chris Lattner, the author of LLVM, Clang and Swift,
+has [a great quote about Kotlin](https://oleb.net/blog/2017/06/chris-lattner-wwdc-swift-panel/).
+I’m gonna put it right here since I deeply agree with him on the subject.
+
+> Swift and Kotlin evolved at about the same point in time with the same contemporary languages around them. And so the surface-level syntax does look very similar. … But if you go one level down below the syntax, the semantics are quite different. Kotlin is very reference semantics, it’s a thin layer on top of Java, and so it perpetuates through a lot of the Javaisms in its model.
+
+> If we had done an analog to that for Objective-C it would be like, everything is an `NSObject` and it’s `objc_msgSend` everywhere, just with parentheses instead of square brackets. And a lot of people would have been happy with that for sure, but that wouldn’t have gotten us the functional features, that wouldn’t have gotten us value semantics, that wouldn’t have gotten us a lot of the safety things that are happening [in Swift].
+
+> I think that Kotlin is a great language. I really mean that. Kotlin is a great language, and they’re doing great things. They’re just under a different set of constraints.
+
+Kotlin hits a nice middle ground between compatibility with the whole Java world and, at the same time,
+a set of features and abilities of a truly modern language. Unfortunately, it brings some
+caveats related to its Java roots and it is too easy to simply forget about them.
+Well, unless you shoot yourself in a foot really badly.
+
```

Comment:
I would add a little summary / conclusion sentence which explicitly mentionesthe „solution“ you propose, something like

> To avoid that, annotate code with `@Nullable` annotations as much as possible and consider using a Facade.